app/tests/heartbeat_controller_desktop_test.cc

```diff
@@ -336,5 +338,181 @@ TEST_F(HeartbeatControllerDesktopTest,
+  std::string day2 = "2000-01-24";
+  // Date provider will be called twice for, Log then twice for Get
+  EXPECT_CALL(mock_date_provider_, GetDate())
+      .Times(4)
+      .WillOnce(Return(day1))
+      .WillRepeatedly(Return(day2));
+
+  controller_.LogHeartbeat();
+  controller_.LogHeartbeat();
+  // GetAndResetStoredHeartbeats is done synchronously, so there is no need to
+  // wait.
+  std::string payload = controller_.DecodeAndDecompress(
+      controller_.GetAndResetTodaysStoredHeartbeats());
+
+  EXPECT_THAT(payload, EqualsJson(R"json({
+      "heartbeats": [
+        {
+          agent: "agent/1",
+          dates: ["2000-01-24"]
+        }
+      ],
+      "version":"2"
+    })json"));
+
+  // The heartbeat for the previous day should still be stored (01-23).
+  std::string payload2 = controller_.DecodeAndDecompress(
+      controller_.GetAndResetStoredHeartbeats());
+
+  EXPECT_THAT(payload2, EqualsJson(R"json({
+      "heartbeats": [
+        {
+          agent: "agent/1",
+          dates: ["2000-01-23"]
+        }
+      ],
+      "version":"2"
+    })json"));
+}
+
+TEST_F(HeartbeatControllerDesktopTest, GetHeartbeatPayloadMultipleTimes) {
+  app_common::RegisterLibrariesFromUserAgent(kDefaultUserAgent);
+  std::string today = "2000-01-23";
+  // Date provider will be called for both Log and Get
+  EXPECT_CALL(mock_date_provider_, GetDate())
+      .Times(3)
+      .WillRepeatedly(Return(today));
+
+  controller_.LogHeartbeat();
+  // GetAndResetStoredHeartbeats is done synchronously, so there is no need to
+  // wait.
+  std::string first_payload = controller_.DecodeAndDecompress(
+      controller_.GetAndResetStoredHeartbeats());
+  EXPECT_THAT(first_payload, EqualsJson(R"json({
+      "heartbeats": [
+        {
+          agent: "agent/1",
+          dates: ["2000-01-23"]
+        }
+      ],
+      "version":"2"
+    })json"));
+
+  std::string second_payload = controller_.DecodeAndDecompress(
+      controller_.GetAndResetStoredHeartbeats());
+  EXPECT_EQ(second_payload, "");
+}
+
+TEST_F(HeartbeatControllerDesktopTest, GetTodaysHeartbeatPayloadMultipleTimes) {
+  app_common::RegisterLibrariesFromUserAgent(kDefaultUserAgent);
+  std::string today = "2000-01-23";
+  // Date provider will be called for both Log and Get
+  EXPECT_CALL(mock_date_provider_, GetDate())
+      .Times(3)
+      .WillRepeatedly(Return(today));
+
+  controller_.LogHeartbeat();
+  // GetAndResetStoredHeartbeats is done synchronously, so there is no need to
+  // wait.
+  std::string first_payload = controller_.DecodeAndDecompress(
+      controller_.GetAndResetTodaysStoredHeartbeats());
+  EXPECT_THAT(first_payload, EqualsJson(R"json({
+      "heartbeats": [
+        {
+          agent: "agent/1",
+          dates: ["2000-01-23"]
+        }
+      ],
+      "version":"2"
+    })json"));
+
+  std::string second_payload = controller_.DecodeAndDecompress(
+      controller_.GetAndResetStoredHeartbeats());
+  EXPECT_EQ(second_payload, "");
+}
+
```

Comment:
I recommend to add the following tests:

1. Dedicated tests against `GetStringPayloadForHeartbeats()` since this function parse the `LoggedHeartbeats` object manually. 
2. Directly test the expected output gzip-archived and BASE64-encoded string similar to what is done in PlatformLoggingParserTest.java, in addition to validating the value after decoding.

Reply:
I added two tests. One calls GetStringPayloadForHeartbeats directly and compares the result to an expected encoded value. The other calls GetAndResetStoredHeartbeats and compares the result to an expected encoded value.

Reply:
In order to unit test GetStringPayloadForHeartbeats, I made the method public. I saw also a macro FRIEND_TEST, but I think all current usages of FRIEND_TEST are actually no-op (guarded by an #ifdef which is false)

Reply:
You can try to use friend class directly

Reply:
Using friend class directly. Each test case is technically a different class, which I guess is what FRIEND_TEST macro tries to simplify, but I'm not sure how to include the gtest_prod header in the cmakelists, so I just call friend class directly.